apis/python/src/tiledbsoma/_tdb_handles.py

```diff
@@ -228,6 +232,55 @@ def _do_initial_reads(self, reader: tiledb.Group) -> None:
         }
 
 
+class _DictMod(enum.Enum):
+    """State machine to keep track of modifications to a dictionary.
+
+    This whole thing is a hack to allow users to treat the metadata dict
+    like an actual dictionary because tiledb currently does not support multiple
+    modifications to the same key (e.g., add-then-delete a metadata entry has
+    undesired results) [sc-25089].
+    """
+
+    ABSENT = enum.auto()
+    """The key is not present in the dict. Initial state."""
+    PRESENT = enum.auto()
+    """The key is in the dict and is unchanged. Initial state."""
+    ADDED = enum.auto()
+    """The key was originally ABSENT but has been added."""
+    UPDATED = enum.auto()
+    """The key was originally PRESENT but has been changed."""
+    DELETED = enum.auto()
+    """The key was originally PRESENT but has been deleted."""
+
+    @classmethod
+    def start_state(cls, dct: Mapping[Any, Any], key: Any) -> "_DictMod":
+        """Returns the starting state for a DictMod given the key of dct."""
+        return cls.PRESENT if key in dct else cls.ABSENT
+
+    def next_state(self, action: Literal["set", "del"]) -> "_DictMod":
+        """Determines the next state of an entry given the action."""
+        return {
+            _DictMod.ABSENT: {
+                "set": _DictMod.ADDED,
+            },
+            _DictMod.PRESENT: {
+                "set": _DictMod.UPDATED,
+                "del": _DictMod.DELETED,
+            },
+            _DictMod.ADDED: {
```

Comment:
nit: could help reader if we group ABSENT/ADDED together, and the other three together, since the two group's cycles are disconnected.